calculator.c

```diff
@@ -470,65 +477,76 @@ int tokenize(char *str, char *(**tokensRef))
 						//printf("Decimal\n");
 						hasDecimal = true;
 						len++;
-						newToken = (char*)malloc((len + 1) * sizeof(char));
-						newToken[0] = '0';
-						newToken[1] = '.';
+						tmpToken[0] = '0';
+						tmpToken[1] = '.';
 					}
 					else // Numbers that do not start with decimal
 					{
-						newToken = (char*)malloc((len + 1) * sizeof(char)); // Leave room for '\0'
-						newToken[len-1] = ch;
+						tmpToken[len-1] = ch;
 					}
 
 					// Assemble rest of number
 					for(; // Don't change len
 						*ptr // There is a next character and it is not null
+						&& len <= MAXTOKENLENGTH 
 						&& (type(*ptr) == digit // The next character is a digit
 						 	|| ((type(*ptr) == decimal // Or the next character is a decimal
 						 		&& hasDecimal == 0)) // But we have not added a decimal
 						 	|| ((*ptr == 'E' || *ptr == 'e') // Or the next character is an exponent
-						 		&& hasExponent == false)); // But we have not added an exponent yet
+						 		&& hasExponent == false) // But we have not added an exponent yet
+						 	|| ((*ptr == '+' || *ptr == '-') && hasExponent == true)); // Exponent with sign
 						++len)
 					{
 						if(type(*ptr) == decimal)
 							hasDecimal = true;
 						else if(*ptr == 'E' || *ptr == 'e')
 							hasExponent = true;
-						newToken = (char*)realloc(newToken, (len + 1) * sizeof(char)); // Leave room for '\0'
-						newToken[len] = *ptr++;
+						tmpToken[len] = *ptr++;
 					}
 
 					// Append null-terminator
-					newToken[len] = '\0';
+					tmpToken[len] = '\0';
 				}
 				break;
 			case text:
 				// Assemble an n-character (plus null-terminator) text token
 				{
 					int len = 1;
-					newToken = (char*)malloc((len + 1) * sizeof(char)); // Leave room for '\0'
-					newToken[0] = ch;
-					for(len = 1; *ptr && type(*ptr) == text; ++len)
+					tmpToken[0] = ch;
+					for(len = 1; *ptr && type(*ptr) == text && len <= MAXTOKENLENGTH; ++len)
 					{
-						newToken = (char*)realloc(newToken, (len + 1) * sizeof(char)); // Leave room for '\0'
-						newToken[len] = *ptr++;
+						tmpToken[len] = *ptr++;
 					}
-					newToken[len] = '\0';
+					tmpToken[len] = '\0';
 				}
 				break;
+			default:
+				break;
 		}
 		// Add to list of tokens
-		if(newToken != NULL)
+		if(tmpToken[0] != '\0')
 		{
 			numTokens++;
 			/*if(tokens == NULL) // First allocation
 				tokens = (char**)malloc(numTokens * sizeof(char*));
 			else*/
-				tokens = (char**)realloc(tokens, numTokens * sizeof(char*));
+			newToken = malloc((strlen(tmpToken)+1) * sizeof(char));
```

Comment:
Since `newToken` isn't needed until all the way down here, let's move its declaration (line 397) here.